improv_ble_client/protocol.py

```diff
@@ -0,0 +1,265 @@
+"""Models for the Improv via BLE protocol."""
+
+from __future__ import annotations
+
+from enum import IntEnum, IntFlag
+import struct
+from typing import Final, TypeVar
+
+from .errors import InvalidCommand
+
+SERVICE_UUID: Final = "00467768-6228-2272-4663-277478268000"
+CHARACTERISTIC_UUID_CAPABILITIES: Final = "00467768-6228-2272-4663-277478268005"
+CHARACTERISTIC_UUID_STATE: Final = "00467768-6228-2272-4663-277478268001"
+CHARACTERISTIC_UUID_ERROR: Final = "00467768-6228-2272-4663-277478268002"
+CHARACTERISTIC_UUID_RPC_COMMAND: Final = "00467768-6228-2272-4663-277478268003"
+CHARACTERISTIC_UUID_RPC_RESULT: Final = "00467768-6228-2272-4663-277478268004"
+
+
+class Capabilities(IntFlag):
+    """State."""
+
+    IDENTIFY = 1
+
+
+class State(IntEnum):
+    """State."""
+
+    AUTHORIZATION_REQUIRED = 1
+    AUTHORIZED = 2
+    PROVISIONING = 3
+    PROVISIONED = 4
+
+
+class Error(IntEnum):
+    """Error."""
+
+    NO_ERROR = 0
+    INVALID_RPC_PACKET = 1
+    UNKNOWN_RPC_COMMAND = 2
+    UNABLE_TO_CONNECT = 3
+    NOT_AUTHORIZED = 4
+    UNKNOWN_ERROR = 0xFF
+
+
+STATE_MAP: dict[str, type[IntEnum | IntFlag]] = {
+    CHARACTERISTIC_UUID_CAPABILITIES: Capabilities,
+    CHARACTERISTIC_UUID_ERROR: Error,
+    CHARACTERISTIC_UUID_STATE: State,
+}
+
+
+class Command:
+    """Base class for commands."""
+
+    cmd_id: int
+    _format: str
+    _len: int
+    _strings: list[bytes]
+
+    def __init__(self, strings: list[bytes]) -> None:
+        """Initialize."""
+        self._format = self._calc_format(strings)
+        self._len = self._calc_len(strings)
+        self._strings = strings
+
+    def as_bytes(self) -> bytes:
+        """Return serialized representation of the command."""
+        return self._pack(self._strings)
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> Command:
+        """Initialize from serialized representation of the command."""
+        cls._validate(data)
+        return cls(cls._extract_strings(data))
+
+    @property
+    def _header(self) -> bytes:
+        """Return packed header."""
+        return struct.pack("!BB", self.cmd_id, self._len)
```

Comment:
Performance won't matter here but you can pre-create the Struct() object if you want instead of doing it every time